Installation issue: NAMD with MPI

Hi,
I would like to build NAMD with MPI support.

### Steps to reproduce the issue
```console
$ git clone https://github.com/spack/spack.git

$ cat >spack/etc/spack/defaults/compilers.yaml <<EOL
compilers:
- compiler:
    spec: gcc@9.2.0
    target: x86_64
    operating_system: centos7
    modules: []
    environment: {}
    extra_rpaths: []
    flags: {}
    paths:
      cc:  /opt/gcc-9.2.0/bin/gcc
      cxx: /opt/gcc-9.2.0/bin/g++
      f77: /opt/gcc-9.2.0/bin/gfortran
      fc:  /opt/gcc-9.2.0/bin/gfortran
EOL

$ cat >>spack/etc/spack/defaults/packages.yaml <<EOL

  openmpi:
    externals:
    - spec: "openmpi @4.0.3 %gcc@9.2.0"
      prefix: /opt/openmpi-4.0.3/
    buildable: False

  intel-mpi:
    externals:
    - spec: "intel-mpi @2019.7.217"
      prefix: /opt/intel/compilers_and_libraries_2020.1.217/linux/mpi
    buildable: False

  intel-mkl:
    externals:
    - spec: "intel-mkl @2019.5.281 threads=openmp"
      prefix: /opt/intel/compilers_and_libraries_2019.5.281/linux/mkl
    buildable: False
EOL

$ source spack/share/spack/setup-env.sh
$ spack install namd ^charmpp backend=mpi ^openmpi
...
<built-in>: note: built-in 'strncmp' declared here
chmod: cannot access '/home/vasco/spack/opt/spack/linux-centos7-zen/gcc-9.2.0/charmpp-6.10.2-gq7vrz5ydfhg2oedstxyuvgjjqaj26dd/bin/../include/MPIOPTS': No such file or directory
Error> charmc can not call AMPI's mpicxx/mpiCC wrapper! Please fix your PATH.
make: *** [inc/ComputeMoaMgr.decl.h] Error 1
make: *** Waiting for unfinished jobs....
Error> charmc can not call AMPI's mpicxx/mpiCC wrapper! Please fix your PATH.
make: *** [inc/ComputeMgr.decl.h] Error 1
Error> charmc can not call AMPI's mpicxx/mpiCC wrapper! Please fix your PATH.
Error> charmc can not call AMPI's mpicxx/mpiCC wrapper! Please fix your PATH.
Error> charmc can not call AMPI's mpicxx/mpiCC wrapper! Please fix your PATH.
make: *** [inc/Node.decl.h] Error 1
make: *** [inc/PatchMgr.decl.h] Error 1
make: *** [inc/WorkDistrib.decl.h] Error 1
Error> charmc can not call AMPI's mpicxx/mpiCC wrapper! Please fix your PATH.
Error> charmc can not call AMPI's mpicxx/mpiCC wrapper! Please fix your PATH.
make: *** [inc/ComputeFmmSerialMgr.decl.h] Error 1
make: *** [inc/ComputeMsmSerialMgr.decl.h] Error 1
Error> charmc can not call AMPI's mpicxx/mpiCC wrapper! Please fix your PATH.
make: *** [inc/ComputeMsmMgr.decl.h] Error 1
```
[spack-build-out.txt](https://github.com/spack/spack/files/5182596/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/5182597/spack-build-env.txt)

Thanks

@matthiasdiener what appears to be happening is that MPICC/MPICXX are being accessed not only when building Charm++ for MPI, but also by charmc when building a program (e.g., NAMD) that **uses** Charm++ built for MPI.

The root of the problem is that Charm++ expects that MPICC/MPICXX will have the same values when building Charm++ and when using it, and it will also attempt to rewrite its include/MPIOPTS file with the changed MPICC/MPICXX values, so setting MPICC=ampicc won't work if Charm++ is built on MPI.

> $ cat ../../../include/MPIOPTS 
MPICXX_DEF=mpicxx
MPICC_DEF=mpicc
$ MPICXX=ampicxx MPICC=ampicc make
../../../bin/charmc   hello.ci
../../../bin/charmc  -c hello.C
../../../bin/charmc: line 207: ampicxx: command not found
Fatal Error by charmc in directory charm-6.10.2/tests/charm++/simplearrayhello
   Command ampicxx -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX -I../../../bin/../include -D__CHARMC__=1 -U_FORTIFY_SOURCE -fno-stack-protector -fno-lifetime-dse -c hello.C -o hello.o returned error code 127
charmc exiting...
make: *** [Makefile:18: hello.o] Error 1
$ cat ../../../include/MPIOPTS 
MPICXX_DEF=ampicxx
MPICC_DEF=ampicc

--- a/var/spack/repos/builtin/packages/charmpp/package.py
+++ b/var/spack/repos/builtin/packages/charmpp/package.py
@@ -370,7 +370,7 @@ class Charmpp(Package):
              'test', 'TESTOPTS=++local', parallel=False)
 
     def setup_dependent_build_environment(self, env, dependent_spec):
-        if not("backend=mpi" in self.spec):
+        if not self.spec.satisfies("backend=mpi"):
             env.set('MPICC',  self.prefix.bin.ampicc)
             env.set('MPICXX', self.prefix.bin.ampicxx)
             env.set('MPIF77', self.prefix.bin.ampif77)

--- a/var/spack/repos/builtin/packages/charmpp/package.py
+++ b/var/spack/repos/builtin/packages/charmpp/package.py
@@ -370,10 +370,11 @@ class Charmpp(Package):
              'test', 'TESTOPTS=++local', parallel=False)
 
     def setup_dependent_build_environment(self, env, dependent_spec):
-        env.set('MPICC',  self.prefix.bin.ampicc)
-        env.set('MPICXX', self.prefix.bin.ampicxx)
-        env.set('MPIF77', self.prefix.bin.ampif77)
-        env.set('MPIF90', self.prefix.bin.ampif90)
+        if not("backend=mpi" in self.spec):
+            env.set('MPICC',  self.prefix.bin.ampicc)
+            env.set('MPICXX', self.prefix.bin.ampicxx)
+            env.set('MPIF77', self.prefix.bin.ampif77)
+            env.set('MPIF90', self.prefix.bin.ampif90)
 
     def setup_dependent_package(self, module, dependent_spec):
         self.spec.mpicc     = self.prefix.bin.ampicc

